## Visualize outputs


In [1]:
#!pip install pytorch-fid

In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
import os,sys,inspect, random
sys.path.insert(0,"..")

import matplotlib.pyplot as plt
from matplotlib import rc
#rc('font',**{'family':'sans-serif','sans-serif':['Computer Modern Roman']})
#rc('text', usetex=True)

import matplotlib.colors
import cv2
import numpy as np


import pytorch_ssim
import torch
from torch.autograd import Variable

In [4]:
def visualize(idx, **images):
    """Plot images in one row."""
    
    #norm=plt.Normalize(0,4) # 5 classes including BG
    #map_name = matplotlib.colors.LinearSegmentedColormap.from_list("", ["black", "red","yellow","blue", "green"])
    
    n = len(images)
    plt.figure(figsize=(10, 8))
    for i, (name, image) in enumerate(images.items()):
        plt.subplot(1, n, i + 1)
        plt.xticks([])
        plt.yticks([])
        #plt.title(' '.join(name.split('_')).title())
        plt.imshow(image)
        
    plt.savefig("../outs/compare/{}.png".format(idx), facecolor="white", bbox_inches = 'tight', dpi=600)
    #plt.show()
    
    
def make_dataset(dir):
    images = []
    assert os.path.isdir(dir), '%s is not a valid directory' % dir

    f = dir.split('/')[-1].split('_')[-1]
    #print (dir, f)
    dirs= os.listdir(dir)
    for img in dirs:

        path = os.path.join(dir, img)
        #print(path)
        images.append(path)
    return images

def read_image(path):
    image = cv2.imread(path)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    return image

In [5]:
os.makedirs('../outs/compare',exist_ok=True)

In [6]:
# Path to model
#algo1 = "acgpn_5000"
#algo2 = "resunet_g1"

# path to tryons

# Same
# cloth = "../results/aligned/TFM/test_pairs/cloth"
# person = "../results/aligned/TFM/test_pairs/person"
# tryon1 = "../results/aligned_unet/TFM/test_pairs/tryon"
# tryon2 = "../results/aligned/TFM/test_pairs/tryon"

# Random
cloth = "../results/random/aligned/TFM/test_pairs/cloth"
person = "../results/random/aligned/TFM/test_pairs/person"
tryon1 = "../results/random/aligned_unet/TFM/test_pairs/tryon"
tryon2 = "../results/random/aligned/TFM/test_pairs/tryon"

#algo1_path = "../tryons_random/" + algo1 + "/" + mode
#algo2_path = "../tryons_random/" + algo2 + "/" + mode

In [7]:
cloths = sorted(make_dataset(cloth))
persons = sorted(make_dataset(person))
tryon1p = sorted(make_dataset(tryon1))
tryon2p = sorted(make_dataset(tryon2))

In [8]:
len(tryon2p)

934

### Visualize different algo output

In [9]:
# slc = 934
# i = 0
# for c, p, a1, a2 in zip(cloths[:slc], persons[:slc], tryon1p[:slc], 
#                      tryon2p[:slc]):
    
#     # visualize(i,
#     #           unet=read_image(a1), 
#     #           resunet = read_image(a2))
#     visualize(i, 
#               person=read_image(p),
#               cloth=read_image(c),
#               unet=read_image(a1), 
#               resunet = read_image(a2))
#     i+=1

In [ ]:
# nums = np.array([4,5,22,45,87,123,10,52,95,
#                  99,323,586,149,188,787,806,829,323,929,325,241,193,276,294,306])

# nums = nums[:12]

nums = [4,5,22,45,87,123,10,52,95,
                 99,323,586,149,188,787,806,829,325,241,306,276,294]

nums = [123, 829, 10, 99, 276, 95, 52, 294] #random.sample(nums, 8)

rows = int(len(nums) / 2)
cols = 8
fig, axarr = plt.subplots(rows, cols, figsize=(22, 18), constrained_layout=True)

v = 0
for r in range(rows):
    rp1=read_image(persons[nums[v+r]])
    tc1=read_image(cloths[nums[v+r]])
    a1=read_image(tryon1p[nums[v+r]])
    o1=read_image(tryon2p[nums[v+r]])
    
    rp2=read_image(persons[nums[v+r+1]])
    tc2=read_image(cloths[nums[v+r+1]])
    a2=read_image(tryon1p[nums[v+r+1]])
    o2=read_image(tryon2p[nums[v+r+1]])
    
    # rp3=read_image(persons[nums[v+r+2]])
    # tc3=read_image(cloths[nums[v+r+2]])
    # a3=read_image(tryon1p[nums[v+r+2]])
    # o3=read_image(tryon2p[nums[v+r+2]])
    
    v+=1
    #images = [rp1, tc1, a1, o1, rp2, tc2, a2, o2, rp3, tc3, a3, o3]
    images = [rp1, tc1, a1, o1, rp2, tc2, a2, o2]
    
    captions = ["Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours", 
                "Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours", 
                "Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours"]
    
    for c in range(cols):
        axarr[r, c].imshow(images[c], cmap='gray')
        axarr[r, c].axis("off")
        axarr[r, c].set_aspect('equal')
        if r==0:
            axarr[r, c].set_title(captions[c], fontsize=25)

plt.savefig("../outs/visualization_tryons_small.pdf", facecolor="white", bbox_inches = 'tight', dpi=100)

In [ ]:
nums

In [ ]:
len(nums)

### All

In [ ]:
nums = np.array([4,5,22,45,87,123,10,52,95,
                 99,323,586,149,188,787,806,829,193,929,325,241,306,276,294])

nums = np.array([4,5,22,45,87,123,10,52,95,
                 99,323,149,188,787,806,829,325,241,306,276,294])

rows = int(7)
cols = 12
fig, axarr = plt.subplots(rows, cols, figsize=(33, 31), constrained_layout=True)

v = 0
for r in range(rows):
    rp1=read_image(persons[nums[v+r]])
    tc1=read_image(cloths[nums[v+r]])
    a1=read_image(tryon1p[nums[v+r]])
    o1=read_image(tryon2p[nums[v+r]])
    
    rp2=read_image(persons[nums[v+r+1]])
    tc2=read_image(cloths[nums[v+r+1]])
    a2=read_image(tryon1p[nums[v+r+1]])
    o2=read_image(tryon2p[nums[v+r+1]])
    
    rp3=read_image(persons[nums[v+r+2]])
    tc3=read_image(cloths[nums[v+r+2]])
    a3=read_image(tryon1p[nums[v+r+2]])
    o3=read_image(tryon2p[nums[v+r+2]])
    
    v+=2
    images = [rp1, tc1, a1, o1, rp2, tc2, a2, o2, rp3, tc3, a3, o3]
    
    captions = ["Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours", 
                "Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours", 
                "Reference \n Person", "Target \n Clothes", "M3D-VTON", "Ours"]
    
    for c in range(cols):
        axarr[r, c].imshow(images[c], cmap='gray')
        axarr[r, c].axis("off")
        axarr[r, c].set_aspect('equal')
        if r==0:
            axarr[r, c].set_title(captions[c], fontsize=35)

plt.savefig("../outs/visualization_tryons_all.pdf", facecolor="white", bbox_inches = 'tight', dpi=100)

### Compute FID scores

In [12]:
!python -m pytorch_fid "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/person/" "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/person/"

100%|███████████████████████████████████████████| 19/19 [00:02<00:00,  8.55it/s]
FID:  -2.020402860125614e-05


#### Baseline

In [13]:
!python -m pytorch_fid "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/person/" "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/tryon"

100%|███████████████████████████████████████████| 19/19 [00:02<00:00,  8.85it/s]
FID:  19.875877419374007


#### Ours

In [14]:
!python -m pytorch_fid "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/person/" "/home/hz/m3d-vton/results/aligned/TFM/test_pairs/tryon"

100%|███████████████████████████████████████████| 19/19 [00:02<00:00,  8.57it/s]
FID:  15.162646143627711


### Compute SSIM

In [16]:
scores = []

gt = "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/person/"
a1 = "/home/hz/m3d-vton/results/aligned_unet/TFM/test_pairs/tryon"
a2 = "/home/hz/m3d-vton/results/aligned/TFM/test_pairs/tryon"

gtp = sorted(make_dataset(gt))
a1p = sorted(make_dataset(a1))
a2p = sorted(make_dataset(a2))

for x,y in zip(gtp, a2p):
    img1 = read_image(x) / 255.0
    img2 = read_image(y) / 255.0
    
    img1 = np.expand_dims(img1, axis=0)
    img2 = np.expand_dims(img2, axis=0)
    
    if torch.cuda.is_available():
        img1 = torch.from_numpy(img1)
        img2 = torch.from_numpy(img2)
        
        #print(img1.shape, img2.shape)
        score = pytorch_ssim.ssim(img1, img2)
        score = score.item()
        #print(score)
        scores.append(score)

In [17]:
score = np.array(scores).mean()
print("SSIM score: {:.4f}".format(score))

SSIM score: 0.9814


| Method | FID (↓) | SSIM (↑)|
|--------|------|-----|
| VITON (CVPR, 2018) | 28.43 | 0.8807 |
| CP-VTON (ECCV, 2018) | 20.05 | 0.8503 |
| CP-VTON+ (CVPRW, 2020) | 23.18 | 0.8782 |
| ACGPN (CVPR, 2020) | 20.19 | 0.8924 |
| M3D-VTON (ICCV, 2021) | 19.875 | 0.9725 |
| Ours | **15.162** | **0.9814** |